Content.Server/Fluids/EntitySystems/PuddleSystem.cs

```diff
@@ -183,7 +227,87 @@ public bool WouldOverflow(EntityUid uid, Solution solution, PuddleComponent? pud
+        }
+
+        /// <summary>
+        ///     Whether adding this solution to this puddle would overflow.
+        /// </summary>
+        /// <param name="uid">Uid of owning entity</param>
+        /// <param name="puddle">Puddle ref param</param>
+        /// <returns></returns>
+        private bool IsOverflowing(EntityUid uid, PuddleComponent? puddle = null)
+        {
+            if (!Resolve(uid, ref puddle))
+                return false;
+
+            return CurrentVolume(uid, puddle) > puddle.OverflowVolume;
+        }
+
+        public PuddleComponent SpawnPuddle(EntityUid srcUid, EntityCoordinates pos, PuddleComponent? srcPuddleComponent = null)
+        {
+            MetaDataComponent? metadata = null;
+            string? prototype = null;
+            if (Resolve(srcUid, ref srcPuddleComponent, ref metadata))
+            {
+                prototype = metadata.EntityPrototype?.ID;
+            }
+
+            var destUid = EntityManager.SpawnEntity(prototype, pos);
+            var destPuddle = EntityManager.EnsureComponent<PuddleComponent>(destUid);
+
+            
+            // if this is metadata-less component we do special puddle spawning
+            if (srcPuddleComponent != null && metadata == null)
+            {
+                // Copy puddle values
+                destPuddle.SlipThreshold = srcPuddleComponent.SlipThreshold;
+                destPuddle.WetFloorEffectThreshold = srcPuddleComponent.WetFloorEffectThreshold;
+                destPuddle.SpillSound = srcPuddleComponent.SpillSound;
+                destPuddle.OverflowVolume = srcPuddleComponent.OverflowVolume;
+                destPuddle.OpacityModifier = srcPuddleComponent.OpacityModifier;
+                destPuddle.SolutionName = srcPuddleComponent.SolutionName;
+                
+                // Copy SolutionContainer properties
+                EntityManager.EnsureComponent<SolutionContainerManagerComponent>(destUid);
+                _solutionContainerSystem.EnsureSolution(destUid, srcPuddleComponent.SolutionName);
+                
+                // Copy visuals
+                var spriteQuery = EntityManager.GetEntityQuery<SpriteComponent>();
+                if (spriteQuery.TryGetComponent(srcUid, out var srcSprite))
+                {
+                    var destSprite = EntityManager.EnsureComponent<SpriteComponent>(destUid);
+                    destSprite.Visible = srcSprite.Visible;
+                    destSprite.DrawDepth = srcSprite.DrawDepth;
+                    destSprite.Scale = srcSprite.Scale;
+                    destSprite.Rotation = srcSprite.Rotation;
+                    destSprite.Offset = srcSprite.Offset;
+                    destSprite.Color = srcSprite.Color;
+                    destSprite.DrawDepth = srcSprite.DrawDepth;
+                    destSprite.RenderOrder = srcSprite.RenderOrder;
+                    destSprite.BaseRSIPath = srcSprite.BaseRSIPath;
+                }
```

Comment:
Don't sprites have an existing API for copying?

Reply:
On client.